File: Ndapi/ProgramUnit.cpp
#include "stdafx.h"

#include <D2FPGU.H>
#include "ProgramUnit.h"

namespace Ndapi
{
	ProgramUnit::ProgramUnit() : NdapiObject()
	{
	}

	ProgramUnit::ProgramUnit(d2fob* program_unit) : NdapiObject(program_unit)
	{
	}

	ProgramUnit::ProgramUnit(FormModule^ form, String^ name)
	{
		Create(name, form, D2FFO_PROG_UNIT);
	}

	void ProgramUnit::Compile()
	{
		auto status = d2fpguco_CompileObj(NdapiContext::Context, _handler);
		CheckStatusAndThrow(status, String::Format("Error compiling the object. Name: {0}, Type: {1}", Name, Type));
	}

	String^ ProgramUnit::Comment::get()
	{
		return GetStringProperty(D2FP_COMMENT);
	}

	void ProgramUnit::Comment::set(String^ value)
	{
		SetStringProperty(D2FP_COMMENT, value);
	}

	String^ ProgramUnit::Text::get()
	{
		return GetStringProperty(D2FP_PGU_TXT);
	}

	void ProgramUnit::Text::set(String^ value)
	{
		SetStringProperty(D2FP_PGU_TXT, value);
	}

	String^ ProgramUnit::Name::get()
	{
		return GetStringProperty(D2FP_NAME);
	}

	ProgramUnitType ProgramUnit::Type::get()
	{
		return safe_cast<ProgramUnitType>(GetNumberProperty(D2FP_PGU_TYP));
	}
}

File: Ndapi/ProgramUnit.cs
using Ndapi.Core;
using Ndapi.Core.Handles;
using Ndapi.Enums;

namespace Ndapi;

/// <summary>
/// Represents a program unit object.
/// </summary>
public partial class ProgramUnit : NdapiObject<ProgramUnit>
{
    /// <summary>
    /// Creates a program unit.
    /// </summary>
    /// <param name="module">Program unit owner.</param>
    /// <param name="name">Program unit name.</param>
    public ProgramUnit(FormModule module, string name) : base(name, ObjectType.ProgramUnit, module)
    {
    }

    /// <summary>
    /// Creates a program unit.
    /// </summary>
    /// <param name="module">Program unit owner.</param>
    /// <param name="name">Program unit name.</param>
    public ProgramUnit(MenuModule module, string name) : base(name, ObjectType.ProgramUnit, module)
    {
    }

    /// <summary>
    /// Creates a program unit.
    /// </summary>
    /// <param name="module">Program unit owner.</param>
    /// <param name="name">Program unit name.</param>
    public ProgramUnit(ObjectGroup module, string name) : base(name, ObjectType.ProgramUnit, module)
    {
    }

    internal ProgramUnit(ObjectSafeHandle handle) : base(handle, ObjectType.ProgramUnit)
    {
    }

    /// <summary>
    /// Gets or sets the comment.
    /// </summary>
    [Property(NdapiConstant.D2FP_COMMENT)]
    public partial string Comment { get; set; }

    /// <summary>
    /// Gets or sets the program unit code.
    /// </summary>
    [Property(NdapiConstant.D2FP_PGU_TXT)]
    public partial string Text { get; set; }

    /// <summary>
    /// Gets the program unit type.
    /// </summary>
    [Property(NdapiConstant.D2FP_PGU_TYP)]
    public partial ProgramUnitType Type { get; }

    /// <summary>
    /// Compile the program unit.
    /// </summary>
    public void Compile()
    {
        var status = NativeMethods.d2fpguco_CompileObj(NdapiContext.GetContext(), Handle);
        Ensure.Success(status);
    }
}
